File: packages/core/src/pomi/extractor.ts
import type { Canary, CanaryEvidence, CanaryAnalysisExactMatch, CanaryAnalysisPattern, CanaryAnalysisStatistical } from '../types.js'

export class CanaryExtractor {
  extract(injectedCanaries: Canary[], canaryResponses: Record<string, string> | undefined): CanaryEvidence[] {
    if (!canaryResponses) return []

    const evidence: CanaryEvidence[] = []

    for (const canary of injectedCanaries) {
      const response = canaryResponses[canary.id]
      if (response === undefined) continue

      const result = this.evaluate(canary, response)
      evidence.push(result)
    }

    return evidence
  }

  private evaluate(canary: Canary, observed: string): CanaryEvidence {
    switch (canary.analysis.type) {
      case 'exact_match':
        return this.evaluateExactMatch(canary, canary.analysis, observed)
      case 'pattern':
        return this.evaluatePattern(canary, canary.analysis, observed)
      case 'statistical':
        return this.evaluateStatistical(canary, canary.analysis, observed)
    }
  }

  private evaluateExactMatch(canary: Canary, analysis: CanaryAnalysisExactMatch, observed: string): CanaryEvidence {
    // Check if observed matches any model's expected value
    let bestMatch = ''
    let match = false

    for (const [_family, expected] of Object.entries(analysis.expected)) {
      if (observed.trim().toLowerCase() === expected.trim().toLowerCase()) {
        bestMatch = expected
        match = true
        break
      }
    }

    // If no match, use the first expected value as reference
    if (!match) {
      bestMatch = Object.values(analysis.expected)[0] ?? ''
    }

    return {
      canary_id: canary.id,
      observed,
      expected: bestMatch,
      match,
      confidence_contribution: match ? canary.confidence_weight : canary.confidence_weight * 0.3,
    }
  }

  private evaluatePattern(canary: Canary, analysis: CanaryAnalysisPattern, observed: string): CanaryEvidence {
    let bestPattern = ''
    let match = false

    for (const [_family, pattern] of Object.entries(analysis.patterns)) {
      try {
        const regex = new RegExp(pattern, 'i')
        if (regex.test(observed)) {
          bestPattern = pattern
          match = true
          break
        }
      } catch {
        // Invalid regex, skip
      }
    }

    if (!match) {
      bestPattern = Object.values(analysis.patterns)[0] ?? ''
    }

    return {
      canary_id: canary.id,
      observed,
      expected: bestPattern,
      match,
      confidence_contribution: match ? canary.confidence_weight : canary.confidence_weight * 0.2,
    }
  }

  private evaluateStatistical(canary: Canary, analysis: CanaryAnalysisStatistical, observed: string): CanaryEvidence {
    // For statistical canaries, we parse the observed value and check if it's
    // within a reasonable range for any model's distribution
    // For now, extract the first number from the response
    const numMatch = observed.match(/-?\d+\.?\d*/)
    const numValue = numMatch ? parseFloat(numMatch[0]) : NaN

    let bestDist = ''
    let match = false

    if (!isNaN(numValue)) {
      for (const [family, dist] of Object.entries(analysis.distributions)) {
        // Within 2 standard deviations
        if (Math.abs(numValue - dist.mean) <= 2 * dist.stddev) {
          bestDist = `${family}: mean=${dist.mean}, stddev=${dist.stddev}`
          match = true
          break
        }
      }
    }

    if (!match) {
      const first = Object.entries(analysis.distributions)[0]
      bestDist = first ? `${first[0]}: mean=${first[1].mean}, stddev=${first[1].stddev}` : ''
    }

    return {
      canary_id: canary.id,
      observed,
      expected: bestDist,
      match,
      confidence_contribution: match ? canary.confidence_weight * 0.7 : canary.confidence_weight * 0.1,
    }
  }
}

File: sdks/python/src/xagentauth/pomi/extractor.py
from __future__ import annotations

import re
from typing import Optional

from xagentauth.types import (
    Canary,
    CanaryAnalysisExactMatch,
    CanaryAnalysisPattern,
    CanaryAnalysisStatistical,
    CanaryEvidence,
)


class CanaryExtractor:
    """Evaluates canary responses to produce evidence for model classification."""

    def extract(
        self,
        injected_canaries: list[Canary],
        canary_responses: Optional[dict[str, str]],
    ) -> list[CanaryEvidence]:
        if not canary_responses:
            return []

        evidence: list[CanaryEvidence] = []
        for canary in injected_canaries:
            response = canary_responses.get(canary.id)
            if response is None:
                continue
            result = self._evaluate(canary, response)
            evidence.append(result)

        return evidence

    def _evaluate(self, canary: Canary, observed: str) -> CanaryEvidence:
        analysis = canary.analysis
        if isinstance(analysis, CanaryAnalysisExactMatch) or (isinstance(analysis, dict) and analysis.get("type") == "exact_match"):
            return self._evaluate_exact_match(canary, analysis, observed)
        elif isinstance(analysis, CanaryAnalysisPattern) or (isinstance(analysis, dict) and analysis.get("type") == "pattern"):
            return self._evaluate_pattern(canary, analysis, observed)
        elif isinstance(analysis, CanaryAnalysisStatistical) or (isinstance(analysis, dict) and analysis.get("type") == "statistical"):
            return self._evaluate_statistical(canary, analysis, observed)
        else:
            raise ValueError(f"Unknown analysis type: {analysis}")

    def _evaluate_exact_match(
        self,
        canary: Canary,
        analysis: CanaryAnalysisExactMatch,
        observed: str,
    ) -> CanaryEvidence:
        best_match = ""
        match = False

        for _family, expected in analysis.expected.items():
            if observed.strip().lower() == expected.strip().lower():
                best_match = expected
                match = True
                break

        if not match:
            values = list(analysis.expected.values())
            best_match = values[0] if values else ""

        return CanaryEvidence(
            canary_id=canary.id,
            observed=observed,
            expected=best_match,
            match=match,
            confidence_contribution=canary.confidence_weight if match else canary.confidence_weight * 0.3,
        )

    def _evaluate_pattern(
        self,
        canary: Canary,
        analysis: CanaryAnalysisPattern,
        observed: str,
    ) -> CanaryEvidence:
        best_pattern = ""
        match = False

        for _family, pattern in analysis.patterns.items():
            try:
                regex = re.compile(pattern, re.IGNORECASE)
                if regex.search(observed):
                    best_pattern = pattern
                    match = True
                    break
            except re.error:
                continue

        if not match:
            values = list(analysis.patterns.values())
            best_pattern = values[0] if values else ""

        return CanaryEvidence(
            canary_id=canary.id,
            observed=observed,
            expected=best_pattern,
            match=match,
            confidence_contribution=canary.confidence_weight if match else canary.confidence_weight * 0.2,
        )

    def _evaluate_statistical(
        self,
        canary: Canary,
        analysis: CanaryAnalysisStatistical,
        observed: str,
    ) -> CanaryEvidence:
        num_match = re.search(r"-?\d+\.?\d*", observed)
        num_value = float(num_match.group(0)) if num_match else float("nan")

        best_dist = ""
        match = False

        if num_value == num_value:  # not NaN
            for family, dist in analysis.distributions.items():
                if abs(num_value - dist.mean) <= 2 * dist.stddev:
                    best_dist = f"{family}: mean={dist.mean}, stddev={dist.stddev}"
                    match = True
                    break

        if not match:
            items = list(analysis.distributions.items())
            if items:
                first_family, first_dist = items[0]
                best_dist = f"{first_family}: mean={first_dist.mean}, stddev={first_dist.stddev}"

        return CanaryEvidence(
            canary_id=canary.id,
            observed=observed,
            expected=best_dist,
            match=match,
            confidence_contribution=canary.confidence_weight * 0.7 if match else canary.confidence_weight * 0.1,
        )
